### PR TITLE
#4 Added cloud implementation

### LINKED ISSUE #4: Add implementation for Exasol on AWS
Add an implementation for Exasol on AWS

### LINKED ISSUE #11: Add AWS CI account setup
Add an setup for an secure AWS CI account

### PR DESCRIPTION
**Pull request by jakobbraun**
Closes #4 
Closes #11 
We extracted the features of this branch from main since they were not yet production ready.
All contents of this branch are already reviewed.

### CODE REVIEW

**Comment by jakobbraun** [comment]
test

File: .github/workflows/ci-build.yml [hunk 1]
```diff
@@ -24,9 +24,28 @@ jobs:
...
+          terraform -chdir=cloudSetup/ apply -auto-approve
+        env:
+          AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          OWNER: ${{ secrets.AWS_TAG_OWNER }}
+          DEPUTY: ${{ secrets.AWS_TAG_DEPUTY }}
       - name: Run tests and build with Maven
-        run: mvn -B clean verify sonar:sonar --file pom.xml -Dsonar.organization=exasol -Dsonar.host.url=https://sonarcloud.io -Dsonar.login=$SONAR_TOKEN
+        run: |
+          . cloudSetup/generated/setEnv.sh
```
Comment (tomuben) [inline .github/workflows/ci-build.yml:40]: Hm, looks strange. Shouldn't be ./cloudSetup/generated/setEnv.sh

File: .github/workflows/ci-build.yml [hunk 2]
```diff
@@ -24,9 +24,28 @@ jobs:
...
+          mvn -B verify sonar:sonar --file pom.xml -Dsonar.organization=exasol -Dsonar.host.url=https://sonarcloud.io -Dsonar.login=$SONAR_TOKEN
         env:
           GITHUB_OAUTH: ${{ secrets.GITHUB_TOKEN }}
           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
-          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
\ No newline at end of file
+          SONAR_TOKEN: ${{ secrets.SONAR_TOKEN }}
+      - name: Destroy cluster
+        run: terraform -chdir=cloudSetup/ destroy -auto-approve
+        if: always()
```
Comment (tomuben) [inline .github/workflows/ci-build.yml:48]: Is this necessary?

File: .github/workflows/quick-checks.yml [hunk 3]
```diff
@@ -0,0 +1,26 @@
...
+          java-version: 11
+      - name: Cache local Maven repository
+        uses: actions/cache@v2
+        with:
+          path: ~/.m2/repository
+          key: ${{ runner.os }}-maven-${{ hashFiles('**/pom.xml') }}
+          restore-keys: |
+            ${{ runner.os }}-maven-
+      - name: Run quick checks
+        run: mvn -B clean verify --file pom.xml -DskipTests
```
Comment (tomuben) [inline .github/workflows/quick-checks.yml:26]: -DskipTests: I assume the tests are disabled intentionally (maybe because you don't have the database?)?

File: README.md [hunk 4]
```diff
@@ -45,5 +45,6 @@ class MyTest {
 
 ## Additional Information
 
+* [Guide for Setting up a Project With Could Resources](doc/user_guide/setup_project_with_cloud_resources.md)
```
Comment (tomuben) [inline README.md:48]:
```suggestion
* [Guide for Setting up a Project With Cloud Resources](doc/user_guide/setup_project_with_cloud_resources.md)
```

File: cloudSetup/variables.tf [hunk 5]
```diff
@@ -0,0 +1,16 @@
+variable "owner" {}
+
+variable "deputy" {
+  default = ""
+}
+
+# Use a short project tag. Long tags will case the exasol cluster creation to fail (see https://github.com/exasol/cloudformation-aws-exasol/issues/3)
```
Comment (tomuben) [inline cloudSetup/variables.tf:7]:
```suggestion
# Use a short project tag. Long tags will cause the exasol cluster creation to fail (see https://github.com/exasol/cloudformation-aws-exasol/issues/3)
```

File: doc/user_guide/setup_project_with_cloud_resources.md [hunk 6]
```diff
@@ -0,0 +1,55 @@
+# Setting up a Project With AWS Resources
+
+In this guide I will explain how to configure a project to run test with resources on AWS.
```
Comment (tomuben) [inline doc/user_guide/setup_project_with_cloud_resources.md:3]:
```suggestion
In this guide I will explain how to configure a project to run tests with resources on AWS.
```

File: doc/user_guide/setup_project_with_cloud_resources.md [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+
+## CI Isolation
+
+First we need to create a secured environment. That includes two aspects:
```
Comment (tomuben) [inline doc/user_guide/setup_project_with_cloud_resources.md:7]:
```suggestion
First we need to create a secure environment. That includes two aspects:
```

File: doc/user_guide/setup_project_with_cloud_resources.md [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+* The user used in CI has minimal permissions (to reduce damage if it's hacked)
```
Comment (tomuben) [inline doc/user_guide/setup_project_with_cloud_resources.md:9]:
```suggestion
* The user created by the CI has minimal permissions (to reduce damage if it's hacked)
```

File: doc/user_guide/setup_project_with_cloud_resources.md [hunk 9]
```diff
@@ -0,0 +1,55 @@
...
+* In case the CI fails to clean up its resources, a scheduled task deletes them (to save us from cost explosion)
+
+We will use the [ci-isolation-aws](https://github.com/exasol/ci-isolation-aws) for this purpose. To set it up copy the [ci-isolation](../../ci-isolation) folder from this repository to your project. Then change the project name in `ci-isolation/src/.../CiIsolationApp.java`.
```
Comment (tomuben) [inline doc/user_guide/setup_project_with_cloud_resources.md:12]:
```suggestion
We will use the [ci-isolation-aws](https://github.com/exasol/ci-isolation-aws) for this purpose. To set it up, copy the [ci-isolation](../../ci-isolation) folder from this repository to your project. Then change the project name in `ci-isolation/src/.../CiIsolationApp.java`.
```

File: doc/user_guide/setup_project_with_cloud_resources.md [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+
+In this file you can also see that we read the required permissions from a json file from the projects resources. This configuration defines the required permissions for starting an Exasol cluster. If you need a different setup or additional resources, just add another file.
+
+When you're done deploy the CI user to AWS using the AWS CDK:
```
Comment (tomuben) [inline doc/user_guide/setup_project_with_cloud_resources.md:16]:
```suggestion
When you're done, deploy the CI user to AWS using the AWS CDK:
```

File: doc/user_guide/setup_project_with_cloud_resources.md [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    #check the setup
+    terraform destroy # IMPORTANT otherwise you'll pay for the running resources
+  ```
+
+## Set Up CI Build
+
+Now copy the [CI build](../../.github/workflows/ci-build.yml) from this project. The build uses the `aws` environment we created earlier. By that it requires an approval before each run. That is very handy since you probably don't want to create the resources for each commit that just fixed a typo but only before merging a pull request.
+
+The CI build creates the setup using Terraform and writes connection details for the Exasol cluster to `cloudSetup/generated/setEnv.sh`. This script sets environment variables with the connection details. In you test you can use the exasol-test-setup-abstraction-java (ETAJ) to create a connection to the Exasol database. The ETAJ will retrieve the credentials from the environment variables. When you do local testing you can create the cluster using Terraform as showed and use it for all your test runs. That will save you a lot of time. To set the credentials run `. cloudSetup/generated/setEnv.sh` in a terminal and then start your tests / IDE in that terminal. That will make the variables available in the tests.
```
Comment (tomuben) [inline doc/user_guide/setup_project_with_cloud_resources.md:55]:
```suggestion
The CI build creates the setup using Terraform and writes connection details for the Exasol cluster to `cloudSetup/generated/setEnv.sh`. This script sets environment variables with the connection details. In your test you can use the exasol-test-setup-abstraction-java (ETAJ) to create a connection to the Exasol database. The ETAJ will retrieve the credentials from the environment variables. When you do local testing you can create the cluster using Terraform as showed above and use it for all your test runs. That will save you a lot of time. To set the credentials run `. cloudSetup/generated/setEnv.sh` in a terminal and then start your tests / IDE in that terminal. That will make the variables available in the tests.
```

File: .github/workflows/quick-checks.yml (hunk 3 above)
Comment (jakobbraun) [reply .github/workflows/quick-checks.yml:26]: yes, here I just run all the other checks as a quick check, since quite often for example project-keeper breaks the build. In those cases it's better if we see it right here, instead of after starting the cluster which took 40 mins.

File: .github/workflows/ci-build.yml (hunk 2 above)
Comment (jakobbraun) [reply .github/workflows/ci-build.yml:48]: yes, otherwise it won't get executed if the previous one filed. And we wan't to destroy the cluster in any case, otherwise it get's costy ;-) .

File: .github/workflows/ci-build.yml (hunk 1 above)
Comment (jakobbraun) [reply .github/workflows/ci-build.yml:40]: the dot says that we import the environment variables from that script. bash readability...

File: src/main/java/com/exasol/exasoltestsetup/SshConnection.java [hunk 12]
```diff
@@ -129,12 +129,12 @@ public SshExecution runCommand(final String command) {
...
      */
     public int findFreePortOnServer() {
-        final String portOutput = runCommand("python3 <<HEREDOC\n" + PORT_FINDER_PYTHON_SCRIPT + "HEREDOC")
+        final String portOutput = runCommand("python <<HEREDOC\n" + PORT_FINDER_PYTHON_SCRIPT + "HEREDOC")
                 .whenFinished().assertExitCodeIsZero().getStdout();
         try {
             return Integer.parseInt(portOutput.trim());
         } catch (final NumberFormatException exception) {
-            throw new IllegalStateException(ExaError.messageBuilder("F-ETAJ-25")
+            throw new IllegalStateException(ExaError.messageBuilder("F-ETAJ-29")
```
Comment (tomuben) [inline src/main/java/com/exasol/exasoltestsetup/SshConnection.java:137]: Is this change intended in this PR?

File: src/main/java/com/exasol/exasoltestsetup/standalone/ConnectionDetailProvider.java [hunk 13]
```diff
@@ -0,0 +1,42 @@
...
+ * This interface provides connection details for running Exasol database.
+ */
+interface ConnectionDetailProvider {
+
+    /**
+     * Get the IP address of the management node.
+     * 
+     * @return ip address
+     */
+    public String getManagementNodeIp();
```
Comment (tomuben) [inline src/main/java/com/exasol/exasoltestsetup/standalone/ConnectionDetailProvider.java:13]: Why not using [INetAddress](https://docs.oracle.com/javase/7/docs/api/java/net/InetAddress.html) instead of String.

File: src/main/java/com/exasol/exasoltestsetup/standalone/ConnectionDetailProvider.java [hunk 14]
```diff
@@ -0,0 +1,42 @@
...
+
+    /**
+     * Get the ip address of a data node.
+     * 
+     * @return ip address
+     */
+    public String getDataNodeIp();
```
Comment (tomuben) [inline src/main/java/com/exasol/exasoltestsetup/standalone/ConnectionDetailProvider.java:20]: same as above

File: src/main/java/com/exasol/exasoltestsetup/standalone/ConnectionDetailProvider.java (hunk 13 above)
Comment (tomuben) [reply src/main/java/com/exasol/exasoltestsetup/standalone/ConnectionDetailProvider.java:13]: Or https://docs.oracle.com/javase/7/docs/api/java/net/InetSocketAddress.html (IP + ssh Port)

File: src/main/java/com/exasol/exasoltestsetup/standalone/StandaloneExasolTestSetup.java [hunk 15]
```diff
@@ -0,0 +1,173 @@
...
+     * Create a new instance of {@link StandaloneExasolTestSetup}.
+     * 
+     * @param connectionDetailProvider provider for the connection details
+     */
+    public StandaloneExasolTestSetup(final ConnectionDetailProvider connectionDetailProvider) {
+        this.connectionDetails = connectionDetailProvider;
+        this.sshConnection = createConfiguredSshConnection();
+        this.dataNodeIds = fetchDataNodeIds();
+        this.localBucketFsPort = this.sshConnection.addForwardPortForwarding(BUCKET_FS_PORT, getADataNode());
+        final int localHttpsPort = this.sshConnection.addForwardPortForwarding(443);
```
Comment (tomuben) [inline src/main/java/com/exasol/exasoltestsetup/standalone/StandaloneExasolTestSetup.java:44]: Magic Number. Better to create a constant

**Review by tomuben — COMMENTED** [review]
Please check my comments

File: src/main/java/com/exasol/exasoltestsetup/SshConnection.java (hunk 12 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/exasoltestsetup/SshConnection.java:137]: yes, it was duplicate

File: src/main/java/com/exasol/exasoltestsetup/standalone/ConnectionDetailProvider.java (hunk 13 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/exasoltestsetup/standalone/ConnectionDetailProvider.java:13]: I also want to support host names.
I renamed the method to make that clear.

File: doc/changes/changes_0.2.0.md [hunk 16]
```diff
@@ -0,0 +1,26 @@
+# exasol-test-setup-abstraction-java 0.2.0, released 2021-08-17
```
Comment (tomuben) [inline doc/changes/changes_0.2.0.md:1]: Don't forget to update the date ;)

File: .github/workflows/ci-build.yml (hunk 2 above)
Comment (tomuben) [reply .github/workflows/ci-build.yml:48]: I see, maybe a comment would be good in this case.

File: .github/workflows/ci-build.yml (hunk 1 above)
Comment (tomuben) [reply .github/workflows/ci-build.yml:40]: maybe a comment would be good then :)

**Review by tomuben — APPROVED** [review]
Approved, but pls consider my comments.

**Review by tomuben — DISMISSED** [review]

File: src/main/java/com/exasol/exasoltestsetup/WaitHelper.java [hunk 17]
```diff
@@ -29,10 +31,19 @@ public static void waitFor(final long milliSeconds) {
...
+     * @param timeoutAfterSec maximum number of seconds to wait
+     * @param jobDescription  name of the condition to wait on. Used in the timeout exception.
      */
-    public static void waitUntil(final BooleanSupplier predicate) {
+    public static void waitUntil(final BooleanSupplier predicate, final int timeoutAfterSec,
+            final String jobDescription) {
+        final long start = System.currentTimeMillis();
         while (!predicate.getAsBoolean()) {
+            final long elapsed = System.currentTimeMillis() - start;
+            if (elapsed > timeoutAfterSec * 1000L) {
```
Comment (tomuben) [inline src/main/java/com/exasol/exasoltestsetup/WaitHelper.java:43]: I usually prefer to use TimeUnit for this kind of conversion...

File: src/main/java/com/exasol/exasoltestsetup/standalone/StandaloneExasolTestSetup.java [hunk 18]
```diff
@@ -51,12 +52,47 @@ public StandaloneExasolTestSetup(final ConnectionDetailProvider connectionDetail
...
+        int tryCounter = 0;
+        do {
+            exaOperationGateway.setBucketPasswords(this.bucketFsReadPassword, this.bucketFsWritePassword);
+            waitFor(100);
+            tryCounter++;
+            if (tryCounter > 10) {
+                throw new IllegalStateException(
+                        ExaError.messageBuilder("E-ETAJ-32").message("Failed to set BucketFS password.").toString());
+            }
+        } while (!this.isBucketFsIsAvailable());
```
Comment (tomuben) [inline src/main/java/com/exasol/exasoltestsetup/standalone/StandaloneExasolTestSetup.java:76]: You could use a for-loop here:
for (int idxTrials=0; idxTrials < 10; idxTrials) {
...
}

Has the advantage that if avoids possible endless loops if anybody changes the code in the future...but you said it eventually will be rewritten in Go, so it's fine to keep as is....

**Review by tomuben — COMMENTED** [review]
Pls check my comments

**Review by tomuben — APPROVED** [review]

File: src/main/java/com/exasol/exasoltestsetup/WaitHelper.java (hunk 17 above)
Comment (jakobbraun) [reply src/main/java/com/exasol/exasoltestsetup/WaitHelper.java]: :+1:

**Review by tomuben — APPROVED** [review]
